### ISSUE TITLE
JetEngine v2.11.0 Beta

### ISSUE THREAD

**Issue by MjHead** [CrocoBlock/suggestions#4595]
Hi,

We are almost getting ready to release JetEngine v2.11.0 and we would like to ask you to help us test JetEngine v2.11.0 Beta.

You can download beta version from your Crocoblock account.

**Please note that is a beta version, not stable release, so its not ready to use on production websites!**

## Relationships functionality re-factoring

We’ve been asked for quite some time to improve Relations functionality. Main users' requests concerned:

- Relations editing interface - make it somewhat similar to the Toolset and ACF
- Expand relations functionality to other WP objects - users, taxonomies, CCTs
- add metadata for relations, etc.

In this version, Relations functionality is completely rewritten from scratch - UI for adding relations, UI for editing related objects, macros for getting lists of related objects.

### Backward Compatibility:
Since the new functionality is drastically different from the current one, there was no way to make full backward compatibility, so the old relations are working in parallel with the new ones. 

Macros, interface for editing of relations settings and UI for adding/removing related items remain fully functional for old relations. And all new features are unavailable for old relations accordingly. User can convert old relations to the new ones - http://prntscr.com/24c2exm

This will automatically transfer all posts, but users will have to manually change macros for relations used.

### Changes in the Relations logic:

The most fundamental change in the logic of relations is that they are now stored in a separate table in the database and are not attached to the post meta. 

Thanks to that, you can now add relations between any types of data in WP.  Besides, thanks to the new code logic, third-party developers can add any relation types they want. Later we'll add a documentation how to do this

### Relations Editing UI:

**General settings** - http://prntscr.com/24c3jnd.

- _Name_ - name of the relation which will be later shown in the macros and dynamic tags settings
- _Parent object_ - the parent object of the relation.
- _Child object_ - child object
- _Relation type_ - type of relation - one to one, one to many, many to many
- _Parent relation_ - parent relation
- _Register controls for parent object_  - add/hide UI to control related items on the parent object edit page (for example, on the editing page of post, user, term, etc.)
- _Allow to create new children from parent_ - this option enables/disables the possibility to create new child properties from parent editing page
- _Register controls for child object_ - add/hide UI to control related items on the child object edit page (for example, on the editing page of post, user, term etc.)
- _Allow to create new parents from children_ - enable/disable this option to create new parent related items from child editing page
- _Register separate DB table_  - registers separate DB table for this specific relations table (otherwise all related items are stored in the same table of relations)

http://prntscr.com/24ca9ut
http://prntscr.com/24cagvm

**Labels** - http://prntscr.com/24c5t5z
- _Parent Object: label of relation box_  - custom label for relate.properties box on the edit page of the parent object (see screenshot below)
- _Parent Object: label of connect button_ - custom label for the relate button on the parent editing page
- _Parent Object: label of select item control_ - custom label for relate option on the parent editing page (the control, not the button)
- _Parent Object: label of create button_ - custom label of relateate button on the parent editing page
- _Child Object: label of relation box_  - custom label for relate entities on the editing page of the child object
- _Child Object: label of connect button_  - custom label for the relate item button on the editing page of the child object
- _Child Object: label of select item control_ - custom label for the looped item control on the editing page of the child object (the control, not the button)
- _Child Object: label of create button_ - custom label for creating a new parent relayed character on the child editing page

If you select CCT as one of the types, you will add another block of options specific to CCT, about this below in a separate section devoted to CCT

**Meta Fields** - http://prntscr.com/24c9duh - this block is in general identical to the same block in the settings of post types, meta boxes and taxonomies, except that not all field types are available to relations due to UI restrictions related to editing these meta fields in the block with related items.

### UI of Related Item Editing:

**Table with a list of all related items** - http://prntscr.com/24caplh
- Add New ... - button for creating a new related item of the desired type. This button is only available if the required control is enabled in the current Relation settings and if the current user has permissions to create the corresponding properties. Clicking this button opens a popup for creating a new control of the desired type. The set of controls may vary depending on the item. You can also add your own controls for any type and their processing from the code.
- Connect ... - button to select item from existing objects of the desired type. When you click on this button, you will get a popup with the control to select the desired byte, if the item has meta fields, you can set these fields in the next step of the same popup.
- Title column - this column will display the name of the corresponding relayed object. Depending on relatable object type you may have different names - for posts and terms - title, for users - logs and e-mail, for custom related items you can set up your own data to be used as title
- Column Meta Data - this column appears only if you added meta fields for the trailers. Click the button to open the popup for editing the meta fields of the main info.
- Column Actions:
-- Edit - opens the default editing page of the relayed item - for post it will be the page of editing the post, etc. The button is only available to users who have access to edit the item, 
-- View - button for switching to the public page of the selected item (if it is available), 
-- Disconnect - button that removes the item from relayed items, 
-- Delete - button that completely removes the selected item (available only to users with appropriate permissions).

### Tools for working with related items on the front:

**_Get Related Items Macro_** - http://prntscr.com/24ce8lj - gets a list of related items from the corresponding relation, of the desired type for the right object. 
**Settings:**
- From Relation - defines the type of relation to get the related items from
- From Object - defines the object to get the related items from - from Child or Parent object. If a Child Object is selected, the IDs of the Child Items for the current Parent Object will be shown (determined by the settings below) and vice versa
- Initial Object ID From - defines the location of the Current Object ID, related objects of which will be searched for. Options:
-- Current Active Object ID (current post - if we're on a single page, current active listing item - if we're inside the listing, etc.), 
-- Current Logged-in User ID, 
-- Current Selected User ID (for example, on the User page in Profile Builder), 
-- Request Variable (from global request or from WP request variables), 
-- Current Object Variable (based on the object defined as current by JetEngine), 
-- Default Current WP Object - always a global current object for the current page defined by WP, as opposed to the 1st option. So on a single page/single post it will always be the current post, even if we are inside the listing added to this page
- Variable Name - if Query Variable or Current Object Variable is selected in the previous option - this option will allow to set the name of the variable, which the plugin will search for in the query or current object

_**Get Related Siblings Macro**_ - outputs contiguously-allocated related items from the same object as the selected one. The settings of this macro are similar to the previous one, the only difference is in the options logic.
- Initial Object ID From - this option outputs the Object ID for which other objects of the same type are needed (if we choose sister items from a Child object, it is the Current Child Object ID), not the Object ID we need to get related objects from. 

_**Relation Meta Data Source for Dynamic Field widget**_ - http://prntscr.com/24chuyp - more complicated case. Allows you to output the meta data of the current Parent/Child pair from the selected relation. Use Cases:

We've output a listing of child related posts on the parent post's page. Now, we want to output metadata that will be specific to the parent/child post pairs in that relation inside the listing. For this, we need to use Relation Meta Data Source and select the right meta field. After that the plugin will define which related object is currently being displayed - parent or child. Using this information, the plugin then will go up the stack of the processed objects on this page and will try to find a corresponding parent object of the desired type. If an object is found, the plugin will retrieve its ID, and using it, as well as the ID of the current child object, will retrieve the metadata. Example https://crorotest.kinsta.cloud/relation-meta/
In this case we have a Listing of Classes, each class can be trained by several coaches and each coach can have several classes. So the schedule of a particular coach for a particular class can only be defined through the metadata of the coach-class pair.

Now let’s consider a more complicated case - suppose we’ve just output a listing of some posts that have some related items. 
For example, I would like to display a Listing of Projects and each project has some Manager (Curator), which is selected from the existing users on the website. And now I would like to show information about the Manager (Curator) of each project in the Listing of Projects. For this, we can use a nested listing and repeat the previous example, but it will significantly increase the load and rendering time of the listing above. In this case we can only use Dynamic Field widget with the right source and the right context - Related Items From Posts by Users http://prntscr.com/24cjhtj This option allows us to replace the current object for this widget with another object and output the information that otherwise would require only a nested listing. So, how did the context work in this case? The context is always bound to a specific relation and the parent/child object pair accordingly. The context takes the current object and determines its position in the relation - parent or child, it uses this information to find the 1st related object for the chosen one and returns it as current. 
How does this work in our example?
We have a relation - Users to Projects. 
Users are the parent type, Projects are the child type. We display a list of certain Projects and want to show information about the User - the Manager (Curator) of this project. 
In this case, the context understands that the Project is currently displayed, so it has to search for the Related User to this project. It locates the ID of this User by the Project ID and uses it to get the User Object. 
Now we can output the metadata for the User/Project pair or just the User's data.

_**Relations Contexts**_ - the logic of the Relations contexts is described above, I will not repeat myself, the only clarification is that these contexts can be used not only for the Dynamic Field widget and the meta fields of relations, but for all widgets, which have the context support.

_**Get Related Items Dynamic Tag,**_ _**Get Related Siblings Dynamic Tag**_ - works similarly to the macro, the only difference is that it can be used in Elementor controls that support such dynamic tags

_**Get Related Grandparents macro**_ - http://prntscr.com/253idcn - Allows you to get "grandparents" for the selected object. 
How does it work? If a parent for the current relation is selected, we can get the related objects ID from that parent relation for the selected child from the current relation. For example, we have a Cities/Countries relation and a parent Countries/Continents relation for it. With the help of this macro we can get the related continent for the current city. For this hierarchy to work properly, we need the parent object of the current relation to be the same as the child item of the parent relation.

### Filters compatibility:
The new relations are filter compatible. 
We can filter listings with child items by parent item IDs and vice versa. 
For example, in the case of Projects with Managers (Curutors), we can filter Projects by Managers (Curators). 
In order to make the filter compatible with the desired relation, you need to specify the correct variable in the filter settings. 
List of all available relations variables registered on the website can be found at the bottom of filter-editing page - http://prntscr.com/24cnrlc.

Also this feature is compatible with SmartFilter indexer (except for CCTs, because of the way CCTs work, CCT item listings cannot be correctly indexed by their related items) Some solution for this problem might appear in future, but at the moment it is not available. It's not a bug or a flaw, it's physically impossible at this point.

### Peculiarities of work with CCT:

As CCT is a completely custom data type, there are certain peculiarities in the work of relations together with CCT, they concern mainly relations and related items editing settings (except for case with indexer).

For CCT you can add a separate block with settings on the relation editing page - http://prntscr.com/24cp1u5. Here you can specify the field of relation to be used as a related item title in the table with the list of related items. As well as a list of fields that will be available when creating a new related item in the popup.

### Forms compatibility:

Complete form compatibility will be added in the 2nd beta version. At the moment, compatibility with Legacy forms from JetEngine is implemented. How does it work?

You can connect an existing related item to another one by using the Connect Relation Items action/notification. In the settings of this action/notification, you can select the Relation itself, the items of which will be connected to each other, the field of the Parent Object ID and the field of the Child Object ID. For example, the Parent Object ID can be stored in a hidden field - it can be Current Post ID, User ID, etc., while the Child object ID can be chosen in some visible control in the form with a list of necessary posts (terms, users, etc.).
We can output the related items from an existing relation using a preset for the exact field - http://prntscr.com/2531lr9. For this we need to do the following in the default value settings of the selected field
specify the relation we want to get data from,
select the object we want to get data for - parent or child,
indicate the location of the Current Object ID, the related items of which we will search for. In other words if we choose the Child object in the previous option, then in this one we must indicate the ID of the Parent object we want to get the child items from and use those child items as field values

Compatibility with JetFormBuilder will work in the same way.

To be added in the next versions:
Possibility to to create related items from a form and edit the relations metadata from the frontend forms.

## Notes

For bugs and remarks, please reply with a comment on this issue here on GitHub.

Please keep this related to the Current beta Only! any other issue or feature request, feel free to open a new issue.

Many thanks for your support and help!

**Comment by rentaterry** [comment]
Good to see a big updates for relations!

**Comment by kochhase** [comment]
Hi thanks for this big update!
how to see other metafields, Listings insite 
https://prnt.sc/24caplh

for now I only see the post title? But it would very helpfull to see all metafields asigned to the CPT. something similar to toolset.
thanks.

![connecting-related-posts](https://user-images.githubusercontent.com/6199062/147833879-1f0ad4d7-088b-474a-a0a3-ebdae784b974.png)
or eaven nicer

![connecting-related-posts _2](https://user-images.githubusercontent.com/6199062/147833900-661386cb-713d-4d14-bf96-d953f5d2401a.png)

**Comment by hicham1989s** [comment]
What will happen with the Meta Box - Post Relations? Like for example attached.

![Screenshot 2021-12-30 130104](https://user-images.githubusercontent.com/96820586/147750247-79592e63-b1c4-4b43-aadb-4f41d1a27ea7.png)
![asdasdasd](https://user-images.githubusercontent.com/96820586/147750305-21057f12-e4d8-4761-884c-d1f6ae692dc2.png)

**Comment by oceandiveloper** [comment]
I'm super excited about this new feature set and I am just getting started testing. I will report back if I run into issues. 

But I already have a related question: What are the plans regarding admin columns to display this new kind of relationships? The callback "jet_engine_custom_cb_related_posts::" doesn't work with this anymore I suppose?
Is there gonna be an integration with Admin Columns Pro that has released a [JetEngine addons](https://www.admincolumns.com/jetengine/) a while ago?

**Comment by oceandiveloper** [comment]
First impressions are good. Couple of remarks:

1) "Child Object: label of relation box" seems to be not applied in child post editor (instead it shows "Parent Object: label of relation box")
![CleanShot 2022-01-01 at 13 43 27@2x](https://user-images.githubusercontent.com/43462095/147844866-cc7229f3-f615-46a0-8c9d-b2d62d5f66c5.jpg)

2) Please make sure that ALL strings can be translated. I'm using LocoTranslate and couldn't find any strings for the strings marked below:
![CleanShot 2022-01-01 at 14 01 30@2x](https://user-images.githubusercontent.com/43462095/147844958-3a31e1d3-8be7-4fca-85c9-aad93b0a11c4.jpg)

3) When allowing to add new child objects in parent object editor and vice-versa, this only is a time saver if I also can fill the meta fields in the modal dialog. Looks like this is not working currently.
![CleanShot 2022-01-01 at 14 06 00@2x](https://user-images.githubusercontent.com/43462095/147845011-37723050-ea60-494f-8339-ef70403f1ff5.jpg)

4) I would have expected to be able to see the relation meta field values in the relation box like @kochhase already mentioned.

I will test a bit more later.

**Comment by tshwanom** [comment]
Edit Profile triggers critical error when connected with cct and Register controls for parent object is Switched on. Please see the video below. I tested with posts and it works
https://nimb.ws/3fgDsY

**Comment by MjHead** [comment]
@kochhase, @oceandiveloper - at the moment you can additional columns into the table or additional fields into the create new item popup only form the code. Here is an example how to do this - https://gist.github.com/MjHead/f215cfbf4e5b1d003d94de63e353d18f

**Comment by MjHead** [comment]
@oceandiveloper - we fixed translation issue and Labels controls. You can download Beta 2 with the fixes from your account. As for callback for the admin columns, yes old callback will not work, we'll investigate this and add some solution until stable release

**Comment by MjHead** [comment]
Hi @tshwanom I can't replicate this with some random CCT and set up like in your video. If you have this bug replicated on some staging website, could you please PM me in the Facebook with access details to check it - https://www.facebook.com/andremjhead/

**Comment by kochhase** [comment]
> @kochhase, @oceandiveloper - at the moment you can additional columns into the table or additional fields into the create new item popup only form the code. Here is an example how to do this - https://gist.github.com/MjHead/f215cfbf4e5b1d003d94de63e353d18f

Hi @MjHead thanks for this interimsolution right now! I have just seen that toolset has changed its backend gui to a similar gui to yours after the 2018 database migration. They have added a quickedit item so that you don't always have to go there to make changes in the linked CPT. 
Unfortunately, the gui before this version (2015-2018) to see above as the second screenshot) was much more elegant. You could edit the data of the linked CPT directly via Ajax. So you could also search for and save data there with listings and taxonomien of the linked CPT. I don't know why Toolset did not continue this. In any case, we have now kept our project https://pruess-oberliga.de/ in the old structure. I would be happy to email you access to the site if you would like to take a look at these functions. This is actually the most elegant backend for larger data projects, which is otherwise only known from large CMS.

**Comment by oceandiveloper** [comment]
I am currently testing the capabilities of the new relations in the frontend and I have some issues to get the dynamic link widget to work so that it shows the title of the parent post. 

Currently with the default value "%title%" in the field "Label", it displays only the title of the current post (not the related post) even though the "Context" is set to "Related Items From..."
I tried the Dynamic Tag "Related Items" for the label but to no avail.

![CleanShot 2022-01-05 at 11 35 30@2x](https://user-images.githubusercontent.com/43462095/148157119-a8248240-71c5-494e-b18f-fa015d0d43bf.jpg)

I uploaded a video to show what my issues is: https://cln.sh/irgMvV
Can you please check whether I'm just setting it up incorrectly or whether there is a bug that needs fixing?

**Comment by oceandiveloper** [comment]
I have another UI suggestion. ACF relations have an option to show the post thumbnail in the search as well as the table of chosen relations. This greatly enhances the look and feel. I think this would be a great addition to make it easier to quickly identify the post you are looking for.

![CleanShot 2022-01-05 at 11 49 37@2x](https://user-images.githubusercontent.com/43462095/148158274-e4208f5d-6d75-4b5e-83b6-81cfed220d0e.jpg)

![CleanShot 2022-01-05 at 11 52 11@2x](https://user-images.githubusercontent.com/43462095/148158285-c86d3bc0-c016-4f65-bdc4-e81c9351da22.jpg)

![CleanShot 2022-01-05 at 11 56 41@2x](https://user-images.githubusercontent.com/43462095/148158473-c612e2e8-6c86-466c-a79f-29c3f123e475.jpg)

Maybe in a future update we could get "table columns" for the relations box similar to the "Admin Columns" of CPTs?

**Comment by oceandiveloper** [comment]
Is there a way to display the count of related items? If my parent post has 5 child posts connected, I'd like to show something like "5 related posts"
![CleanShot 2022-01-05 at 12 22 53@2x](https://user-images.githubusercontent.com/43462095/148160388-e9852885-9600-4672-b5e8-ec455d8b38b4.jpg)

**Comment by moxet** [comment]
@MjHead is there any updates in next version to include list of users in select box form? It will open wide rang of functionalities for us.

**Comment by OTW-DEV-1** [comment]
SWEEET, been waiting for this for a while!

i just installed and i testing some things. Not sure if I don't undesrtand or if it dosnt work yet:

I have a CPT and have created a relationship CPT > USER Many to Many

2 things: 
1. I can't get the column to show the user in the post list
2. I tried to create a search filter. I have a listing with the CPTs and I want to search based on related user "name" .. Dosnt seem to work. 

looking forward to your reply and keep up the great work ;)

**Comment by Lonsdale201** [comment]
> Is there a way to display the count of related items? If my parent post has 5 child posts connected, I'd like to show something like "5 related posts" ![CleanShot 2022-01-05 at 12 22 53@2x](https://user-images.githubusercontent.com/43462095/148160388-e9852885-9600-4672-b5e8-ec455d8b38b4.jpg)

You should try this:
![image](https://user-images.githubusercontent.com/23199033/148430526-0d1cc0db-10f2-487e-82de-154b7cb7e189.png)

**Comment by oceandiveloper** [comment]
Another issue: I tried to use JetFormBuilder for post creation (not using anything for relations) and upon submission I directly hit an uncaught error that indicates something with JetEngine Relations.

After reverting JetEngine back to 2.10.3 I was able to submit the form and the post was created successfully. With 2.11.0-beta2 I have this problem. I can also open a support ticket if that's helping.

![CleanShot 2022-01-10 at 12 21 27@2x](https://user-images.githubusercontent.com/43462095/148718500-da8c4b6e-75db-4dde-b97d-9be5bcd594f2.jpg)

**Comment by MjHead** [comment]
@oceandiveloper Thanks for the reporting, we fixed the issue, fix will be released with next Beta

**Comment by MjHead** [comment]
Hi @moxet Sorry for the delay. You mean use list of users as source of options for Select, Radio and Checkboxes fields in the JetFormBuilder?

**Comment by arekuki** [comment]
Will there be?
Social Login in the widgets Login Form and Registration Form
- Google
- Facebook

**Comment by moxet** [comment]
> Hi @moxet Sorry for the delay. You mean use list of users as source of options for Select, Radio and Checkboxes fields in the JetFormBuilder?

Exactly! this is what i am looking for, we can create an internal messaging and notification system in crocoblock if got list of users as select, right now we have posts, term and glossaries etc.

**Comment by Lonsdale201** [comment]
Yes, it would open up new possibilities, so let's hope it gets in. :)

**Comment by Lonsdale201** [comment]
Little bug:

**In the relations labels settings:**
https://prnt.sc/26dyilw Fill the label, and save it. -> https://prnt.sc/26dyj0r everything correct. Now delete the label -> https://prnt.sc/26dyj96 and update the relation. New result -> https://prnt.sc/26dyjl2
Looks like Does not reset to default text.

**Comment by MjHead** [comment]
> Little bug:
> 
> **In the relations labels settings:** https://prnt.sc/26dyilw Fill the label, and save it. -> https://prnt.sc/26dyj0r everything correct. Now delete the label -> https://prnt.sc/26dyj96 and update the relation. New result -> https://prnt.sc/26dyjl2 Looks like Does not reset to default text.

Hi! Thanks we fixed this issue and released RC1 with all latest bug fixes and improvements

**Comment by kochhase** [comment]
> @kochhase, @oceandiveloper - at the moment you can additional columns into the table or additional fields into the create new item popup only form the code. Here is an example how to do this - https://gist.github.com/MjHead/f215cfbf4e5b1d003d94de63e353d18f

Hi Mjhead. in RC2 there is no additional columns to see if we use the admin tool of the CPT. When we will see that?

**Comment by jakobmayerhofer** [comment]
> Hi @moxet Sorry for the delay. You mean use list of users as source of options for Select, Radio and Checkboxes fields in the JetFormBuilder?

This would be wonderful for me as well. It will give us a huge amount of possibilities, to create online platforms, where for example vendors can offer products or services and assign rights to use them to customers. 

I am currently working on a project where I would need to be able to assign users to certain tasks via the front end. If there was a functionality, that would allow me to select a user and assign it to a CPT (eather via relations or via the author metafield) using a form, that would be amazing :)
